### ISSUE TITLE
mediafire click stealing redirects not blocked

### ISSUE THREAD

**Issue by JonasNo** [schomery/popup-blocker#12]
URL: http://www.mediafire.com/file/vbd2dwas2rckdj5/firadisk-src-0.0.1.30.7z
Click the green Download button.

Sometimes you'll see popup-blocker message but you have no time to click on it before redirection.
Other times redirects instantly.

You have to repeat: click Download button, get redirected, click browser back button. Several times before you get the actual file to download.

Side note:
All of these are confusing. The descriptions don't help.
White-list (popup)
White-list (webpage)
Black-list (webpage)
Description of 'White-list (webpage)' and 'Black-list (webpage)' makes it seem like they do the same thing.

**Comment by schomery** [comment]
> All of these are confusing. The descriptions don't help

You are welcome to provide suggestions

**Comment by schomery** [comment]
> Sometimes you'll see popup-blocker message but you have no time to click on it before redirection.

Well, sometimes websites check the popup appearance and when they detect it is unsuccessful, they perform page redirection to the distinction advertiser website. There is no easy way to prevent this.

I added an option in the options page to prevent page redirection for 10 seconds after a popup is requested. This way if page wants to redirect exactly after an unsuccessful popup requests, you will be warned to allow or disallow this redirect request.

This option is not active by default as the page is going to display a confirm message and this might be annoying to users. 

you can give the dev version a try and let me know if it helps or not. Instruction to install the dev version: https://github.com/schomery/popup-blocker/issues/8#issuecomment-271232524

@narcolepticinsomniac this new option might be of your interest too.

**Comment by narcolepticinsomniac** [comment]
Yes, this option is definitely interesting to me. Unfortunately, it isn't working like I'd expect. I tested on the above link: http://www.mediafire.com/file/vbd2dwas2rckdj5/firadisk-src-0.0.1.30.7z

and also on one of the previous problem sites who appear to use a similar method:
http://thevideo.me/74257aar9ymg

In neither case did it catch the redirect. They happen pretty much instantly so maybe it's a speed issue. If it was working, I'd think 10 seconds would be overkill. I don't think I've seen any with much of a delay.

It may be worth noting that when I went to delete storage for mediafire when testing, they are using something like "Fastpopsnumsexpire" in local storage. It won't be in settings until you've satisfied their popup limit (at least 3 or 4 in this case), but you can probably see it in devtools before that. I've seen it before and manipulated its expiration to get rid of them in the past.

**Comment by schomery** [comment]
@narcolepticinsomniac should have been fixed.

**Comment by narcolepticinsomniac** [comment]
Yes, it is somewhat working now. This is a great concept, but for whatever reason it isn't working on the malicious redirects. For example:

http://vidup.me/pm656l4es50q

If you click "Proceed to video", a popup is blocked, the pageload is caught, and it asks for confirmation. Once on the video page, click anywhere and a popup is blocked, but the redirect is not caught. It also still isn't working for the above mediafire link. 

These types of redirects are nasty and becoming more commonplace, so this would be a useful feature, but I'd also imagine there being a 100 to 1 ratio of false positives. If you ever do get it working, maybe it could be for a specified blacklist.

**Comment by narcolepticinsomniac** [comment]
This new feature also doesn't seem to disable correctly when you uncheck the option and save.

**Comment by schomery** [comment]
Is this fixed?

> maybe it could be for a specified blacklist.

I reduced the timer to 2 seconds. I think if it works, it would not be very annoying anymore.

**Comment by narcolepticinsomniac** [comment]
Still buggy. Mediafire and Vidup both opened popups and redirected, which would be a step backwards. The redirect from Mediafire was blocked by Bitdefender whereas the Vidup redirect disappeared on its own after a few seconds (that may be Ublock kicking in). Vidup was still a never-ending loop of the landing page though, so that'd be useless.

It also still won't disable correctly once it has been enabled. 

I made a screen record, pay no attention to the external application window, it's just my download manager.
[New Feature.zip](https://github.com/schomery/popup-blocker/files/717666/New.Feature.zip)

**Comment by schomery** [comment]
It seems the popup blocker is broken in this screen record. Can you give it a try on a clean profile on this page http://tools.add0n.com/popup-blocker.html

If that option is checked, test 14 should be passed (popup is blocked and there is no page redirection)

**Comment by narcolepticinsomniac** [comment]
Yeah, for whatever reason that install had become borked. Yes, it works on the test. I think this is somewhat another case of an extension conflict. On the Mediafire test page:

http://www.mediafire.com/file/vbd2dwas2rckdj5/firadisk-src-0.0.1.30.7z

If you click the download button with just PBS installed there is no redirect to catch. The site doesn't appear to be detecting whether the popup opens and then redirecting if it doesn't. 

If you have PBS and ABP installed with only Easylist enabled (very common), the popup is blocked and the page is redirected. Weird, because even if the popup being blocked isn't exactly the cause of the redirect, it still occurs directly after, so you'd thing it'd still prevent it.

**Comment by schomery** [comment]
The script on mediafire.com is very complex. Seems like it creates a Flash object element and clicks on a link inside this object. This method seems to be able to bypass `window.onbeforeunload` listener. I might be mistaken though.

**Comment by narcolepticinsomniac** [comment]
I always have flash disabled completely, so I'm not sure how they'd pull that off. Maybe I'm misunderstanding your explanation.

**Comment by schomery** [comment]
Maybe it is not flash then! Still I am not sure how they actually perform redirect. Maybe by injecting a PDF and clicking a link over that PDF (there was an embedded PDF creating somewhere in their code). This script is complex and I wasn't able to find the logic.

**Comment by narcolepticinsomniac** [comment]
Not PDF either, it's also disabled. The last year or so, these horrible new methods they've started using to initiate popups have started throwing PDF browser notifications when blocked. I disabled the PDF plugin as soon as it started.

**Comment by schomery** [comment]
Okay, give the new version a try

**Comment by narcolepticinsomniac** [comment]
Huge improvement. Really nice work! It's definitely catching the redirects in the examples now. 

I have a couple suggestions that might really improve this feature if they're possible. First off, it should ignore redirects to the same domain. This would eliminate the majority of false positives.

If it ignored the same domain, the browser notification could be more informative, like: 
"Popup Blocker (strict) detected a non-popup redirect to a different domain
Decide whether to "Leave" or "Stay""

It also doesn't prevent the redirect from attempting to execute over and over again. Browser notifications have a feature to prevent subsequent warnings. If you utilize that, you could also add the dialogue: 
"To apply this decision to any similar redirects, check "Prevent this site from opening browser notifications" or whatever the terminology.

In the Mediafire example, redirects are caught, but the malicious redirect and the href to the actual download occur simultaneously. It's a never-ending loop of the notification and no download if you keep clicking the button. I'm not sure if allowing same domain would be enough to allow one and not the other, or if the notification would still kick in. If both occur simultaneously, the "same domain" should be allowed and the other ignored completely.

When I reference "same domain", perhaps this should also apply to different domains that are obviously "downloads" which are the intended destinations. The "same domain" whitelist exceptions could also be applied to URLs with common file extensions, since sometimes websites store downloads on separate servers.

This all makes sense in my head, but that doesn't mean it'd all be simple to implement. IMO, these tweaks would make an already very cool feature, even more useful.

**Comment by Keith94** [comment]
>First off, it should ignore redirects to the same domain.

Isn't this already covered by the option "Do not block popup requests from same origin"? Correct me if I'm wrong.

**Comment by narcolepticinsomniac** [comment]
Nah. Not the same thing. Selecting "Do not block popup requests from same origin" affects regular popup blocking capabilities negatively, but in the case of same page redirects, it would be very useful.

**Comment by schomery** [comment]
> it should ignore redirects to the same domain.

nice one, unfortunetly not possible to be implemented. Well, not optimally. Currently we install an unload listener and this unload listener is called before unload takes place. For security reasons there is no information about the new destination and hence not possible to filter those redirections.

Although it is possible to have a WebRequest inside background tab, but it is not really a proper way. These global WebRequest observers have many performance issues.

btw, for the same security reason, it is not possible to change the before unload message.

**Comment by narcolepticinsomniac** [comment]
> Although it is possible to have a WebRequest inside background tab, but it is not really a proper way. These global WebRequest observers have many performance issues.

I noted that these same-page redirects were becoming more common, and they've only become more so in recent months. This PBs feature really sets it apart from others. They obviously don't care about "proper ways" to write scripts, so if there is any hackish solution, I would urge you to at least test it out.

If the feature allowed "same domain redirects", it could probably lose the browser alerts, and be enabled by default, making the whole process automated and less confusing as to whether the redirect is actually malicious.

**Comment by schomery** [comment]
I am working on this...

**Comment by narcolepticinsomniac** [comment]
@schomery check your email (unrelated stalking).

**Comment by narcolepticinsomniac** [comment]
Any progress?

**Comment by schomery** [comment]
This is a simple extension that blocks all the redirects if host names do not match. This is a very simple and untested extension. Honestly, I forgot to update this thread when I wrote it while back. Give it a try and let me know how it performs as expected! Don't forget to disable PB's page redirect while testing this.

[redirect.zip](https://github.com/schomery/popup-blocker/files/1145779/redirect.zip)

**Comment by narcolepticinsomniac** [comment]
If it's meant to test the hack to ignore same domain, it's a success for the most part. As a standalone, it's a blunt instrument, and probably more annoying than useful. The only way it'll be worthwhile, is if it can be incorporated into PBs's current implementation, where it only kicks in if initiated directly after a blocked popup.

I did encounter examples where the redirect was a same-domain/subdomain which was blocked. Not sure how that should work.

![mediafire](https://user-images.githubusercontent.com/14100003/28192562-0790deb0-6805-11e7-8948-9874a9319c26.png)

![github](https://user-images.githubusercontent.com/14100003/28192716-fcd5f3ba-6805-11e7-8287-982c6ed92fd5.png)

Regardless of whether same-domain/subdomain can be fine-tuned, it'd probably be good to exclude URLs which are clearly downloads (zip, 7zip, mp4, etc.).


Also, there's a weird issue with links opened in a new tab. You can right-click/open in a new tab your redirect zip above. It opens a blank page when the redirect standalone is enabled.
